internal/destination/destination.go

```diff
@@ -150,6 +142,27 @@ func (d *Destination) Open(ctx context.Context) error {
 		return fmt.Errorf("init jetstream writer: %w", err)
 	}
 
+	// Async handlers & callbacks
+	conn.SetErrorHandler(internal.ErrorHandlerCallback(ctx))
+	conn.SetDisconnectErrHandler(internal.DisconnectErrCallback(ctx, func(c *nats.Conn) {
+		if err := d.writer.Close(); err != nil {
+			sdk.Logger(ctx).
+				Error().
+				Err(err).
+				Msg("no able to close the destination writer (disconnect)")
+		}
+	}))
+	conn.SetReconnectHandler(internal.ReconnectCallback(ctx, func(c *nats.Conn) {
+		if err := d.Open(ctx); err != nil {
+			sdk.Logger(ctx).
+				Error().
+				Err(err).
+				Msg("no able to overwrite the destination writer (reconnect)")
+		}
+	}))
```

Comment:
So while we are reconnecting it's possible that a new record comes in and the destination tries to write it. At the same time, the reconnect logic will overwrite the `writer` field. Won't this cause a race condition? Or am I misunderstanding something?

Reply:
With these changes, we moved to pull mode instead of push mode. NATS no longer pushes messages to the connector; it pushes to a consumer queue from which the connector will pull.

Reference: https://docs.nats.io/nats-concepts/jetstream/consumers

Reply:
Thanks for the clarification of that part. What I meant is: while we are reconnecting, it's possible that the destination connector receives a record for writing.

Reply:
If the rewrite fails, the message will not be acknowledged, and NATS will re-emit eventually, right?

Reply:
Another alternative would be to have one connection for the source & destination. What do you think?

Reply:
It's possible we're not on the same page... My concern was this field, [Destination.writer](https://github.com/conduitio-labs/conduit-connector-nats-jetstream/blob/6ab431b27502a45c612fb1d7e18bbe809e6a8d47/internal/destination/destination.go#L33).

Let's assume this: `Destination.Write()` is called. It will use `Destination.writer` to write the record.

At the same time, we try to reconnect to NATS. The reconnection code will try to overwrite the `Destination.writer` field, which is at the same time being used to write a record.

Reply:
I see what you are saying. The `Open` is also not recreating the jetstream context. I will remove the Destination changes from this PR and make another one for those.

Reply:
Removed.

Reply:
:+1: